topdown/builtins/builtins.go

```diff
@@ -28,6 +29,76 @@ func (c Cache) Get(k interface{}) (interface{}, bool) {
 	return v, ok
 }
 
+type NDBCache map[string]map[ast.Value]ast.Value
```

Comment:
The reason for this new type was mostly to make serdes to/from JSON easier. This type is also more precise than a plain `builtins.Cache`, which greatly reduces the chance that someone could accidentally put a value into it that could not be handled properly.

It also forcibly organizes the input/output values from the ND builtins (which would be difficult to disambiguate later otherwise!) Originally, the cache values from `time.now_ns` would be in the cache, right next to values cached for `http.send`, with no special markings to tell them apart beyond different argument lists. Now, all of the builtins are separated out under their names, preventing any chance of accidental cache collisions if they happen to have the same argument types list.

Reply:
This type wound up being inadequate: I had forgotten that Go has strict rules about how it compares `map` keys, and we rely on custom value equality functions for checking if two `ast.Value` types are equivalent. This meant that no matter what, the ND builtins cache would *never* be used by builtins, because the `ast.Value`s being used to check the map would never match the pointers of the original `ast.Value`s that were used as keys.

The fix was switching to using a `map[string]ast.Object`, because our `Object` type understands how to compare `ast.Value` types for equality. It's a lot of extra machinery unfortunately, but it works well.